Modernize packaging a bit

bluetech opened the pull request:
Adding pyproject.toml allows the package to be installed using PEP517 instead of legacy `setup.py install` method. At least pip says so in a warning when installing without the `wheel` package available.

Change setup.py to setup.cfg (using a script[0]) because setup.py is out of fashion these days.

Remove MANIFEST.ini since LICENSE is implied[1].

[0] https://github.com/gvalkov/setuptools-py2cfg
[1] https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-license-files

JelleZijlstra commented on setup.cfg:
```diff
@@ -1,2 +1,32 @@
-[bdist_wheel]
-universal=0
+[metadata]
```
Might as well put this config in pyproject.toml?

bluetech replied:
I don't have python 3.5 installed, but I suspect it won't work with it. I'll try getting py3.5 and check it though.

hauntsaninja replied:
We can drop 3.5, 3.6 and 3.7 support

bluetech replied:
@hauntsaninja great, that might help. I'll send a PR to do this first.

bluetech commented:
Changed from setup.cfg to pyproject.toml, seems to work. Added it in a separate commit just in case.

~~One final change can be done to eliminate the `tool.setuptools` `py-modules` section by moving from a single module `mypy_extensions.py` to a package, then setuptools can auto-discover it.~~ Actually seems to work with the module as well, so just removed it.

cdce8p commented on pyproject.toml:
```diff
@@ -0,0 +1,29 @@
+name = "mypy_extensions"
+version = "1.0.0-dev"
+description = "Type system extensions for programs checked with the mypy type checker."
+readme = "README.md"
```
That's not necessary. Setuptools can detect that it's a markdown file and adds `Description-Content-Type: text/markdown` to the core metadata.

cdce8p commented on pyproject.toml:
```diff
@@ -0,0 +1,29 @@
+version = "1.0.0-dev"
+description = "Type system extensions for programs checked with the mypy type checker."
+readme = "README.md"
+license = {file = "LICENSE"}
```
```suggestion
license = {text = "MIT"}
```
I'd suggest to use the SPDX license string here instead. The license file is already included by setuptools.

cdce8p commented on pyproject.toml:
```diff
@@ -0,0 +1,29 @@
+    "Programming Language :: Python :: 3.8",
+    "Programming Language :: Python :: 3.9",
+    "Programming Language :: Python :: 3.10",
+    "Programming Language :: Python :: 3.11",
+    "Topic :: Software Development",
```
```suggestion
    "Programming Language :: Python :: 3.11",
    "Programming Language :: Python :: 3.12",
    "Topic :: Software Development",
```
Support for `3.12` was declared in https://github.com/python/mypy_extensions/commit/9dd6d98da338a3924b0ed7440b843988cd5805df

cdce8p reviewed (changes requested):
> ~One final change can be done to eliminate the `tool.setuptools` `py-modules` section by moving from a single module `mypy_extensions.py` to a package, then setuptools can auto-discover it.~ Actually seems to work with the module as well, so just removed it.

I'd keep it. Explicit is often better than implicit. I.e. add
```toml
[tool.setuptools]
py-modules = ["mypy_extensions"]
```

cdce8p commented:
The changes here have been merged as part of https://github.com/python/mypy_extensions/commit/6d9c7b756486a654e795095d393bf8206cd11dea. Thanks a lot @bluetech!
Guess this one can be closed now.

hauntsaninja commented:
Thank you!
